### ISSUE TITLE
Learning rewards using graph neural networks

### ISSUE THREAD

**Issue by akpas001** [pyg-team/pytorch_geometric#3457]
### Discussed in https://github.com/pyg-team/pytorch_geometric/discussions/3455

<div type='discussions-op-text'>

<sup>Originally posted by **akpas001** November  8, 2021</sup>
I need to train a Gnn based on the data of pre existing next_states and rewards. next states are one hot encoded and rewards are random numbers(can be negative as well as float). I have constructed a gnn using GINConv.

According to Markov's decision process in Reinforcement learning, a state_action pair as an input to the network gives the next_state and reward.  (state, action, next_states are one-hot encoded).

So, i am forming a state action pair inside the network and trying to calculate next state and reward loss. the loss graph for next states is good but the graph for rewards is very bad(rewards are not learning). I donot understand what mistake i am doing here. Can somebody help me?

Here I am attaching the snippets of the code, and graphs of next_state loss and reward loss respectively. Can somebody help me with that? 

![1](https://user-images.githubusercontent.com/56883247/140769475-d950a306-9e82-4486-a886-d1611d147c80.PNG)
![2](https://user-images.githubusercontent.com/56883247/140769479-79a5fb26-b531-470c-ad9d-275b716722a1.PNG)
![3](https://user-images.githubusercontent.com/56883247/140769481-e98755f5-81ba-4dec-aa06-bb2b11e0e657.PNG)
![4](https://user-images.githubusercontent.com/56883247/140769483-47bcec0e-ec11-4818-9f9e-ba1c5ea3d963.PNG)
</div>

**Comment by akpas001** [comment]
So, what i am trying to do with action for loop is, 

I am basically trying to append actions to node features to every node, (i have 9 nodes in total), so my x basically is (batch_size*9,30). (num_features = 30), and i am reconstructing it to(batch_size*9,10), then appending actions to it which is again making it (batch_size*9,40).

**Comment by rusty1s** [comment]
So the shape of `x` is `[batch_size, num_nodes, num_features]` or `[batch_size * num_nodes, num_features]`? In case of the latter, you will also need to replicate `edge_index`, as it currently only points to indices to the first 9 nodes.

**Comment by akpas001** [comment]
The latter. I processed the data through dataloader, so all of that had been taken care of.

**Comment by rusty1s** [comment]
How does the final `act` look like? Is that a one-hot encoded vector?

**Comment by akpas001** [comment]
yes, it is a one hot encoded vector. I am attaching an image for your reference. Size is (288,30) because, batch size = 32 and num_nodes = 9. now this gets appended along dim = 1 for every node.
![1](https://user-images.githubusercontent.com/56883247/141340294-fe7e35bf-7db2-41dc-b03e-96e5b092e76b.PNG)

**Comment by rusty1s** [comment]
Thanks. One thing I don't understand in your network is the `flatten()` call. Shouldn't this be a `global_mean_pool`?

**Comment by akpas001** [comment]
even i'm not sure about that, I thought of trying both.

**Comment by akpas001** [comment]
but the network is giving a loss of 0.3 only... i still do not understand the errors around it

**Comment by rusty1s** [comment]
How does your new network look like? In the current one, you have a `Linear` layer with input channels `batch_size * 360`, which is quite weird as the number of parameters should never depend on the batch size :)

**Comment by akpas001** [comment]
I don't think that is an issue because I am using fc1 for getting one of the outputs. and the loss doesn't improve even in that case.

**Comment by akpas001** [comment]
Regarding the network i am using almost the same network, except I rewrote the network into a GraphAutoEncoder.

**Comment by rusty1s** [comment]
You are right. As far as I can see, there isn't a mistake in computing `x1` although the computation of `x2` might be still worth to fix. I'm sorry that I cannot be of more help :(

**Comment by akpas001** [comment]
Thanks for the help!
